elements/lisk-chain/src/state_store/state_store.ts

```diff
@@ -12,78 +12,231 @@
-		deleted: [],
-	};
-
-	for (const diff of stateDiffs) {
-		diffToEncode.updated = diffToEncode.updated.concat(diff.updated);
-		diffToEncode.created = diffToEncode.created.concat(diff.created);
-		diffToEncode.deleted = diffToEncode.deleted.concat(diff.deleted);
-	}
+export interface KeyValue {
+	key: Buffer;
+	value: Buffer;
+}
 
-	const encodedDiff = codec.encode(stateDiffSchema, diffToEncode);
-	batch.put(concatDBKeys(DB_KEY_DIFF_STATE, formatInt(height)), encodedDiff);
-};
+export interface DecodedKeyValue<T> {
+	key: Buffer;
+	value: T;
+}
 
 export class StateStore {
-	public readonly account: AccountStore;
-	public readonly chain: ChainStateStore;
-	public readonly consensus: ConsensusStateStore;
+	private readonly _db: DatabaseReader;
+	private readonly _prefix: Buffer;
+	private _cache: CacheDB;
+	private _snapshot: CacheDB | undefined;
+
+	public constructor(db: DatabaseReader, prefix?: Buffer, cache?: CacheDB) {
+		this._db = db;
+		this._prefix = prefix ?? DB_PREFIX_STATE_STORE;
+		this._cache = cache ?? new CacheDB();
+	}
+
+	public getStore(moduleID: number, storePrefix: number): StateStore {
+		const moduleIDBuffer = Buffer.alloc(4);
+		moduleIDBuffer.writeInt32BE(moduleID, 0);
+		const storePrefixBuffer = Buffer.alloc(2);
+		storePrefixBuffer.writeUInt16BE(storePrefix, 0);
+		const subStore = new StateStore(
+			this._db,
+			Buffer.concat([DB_PREFIX_STATE_STORE, moduleIDBuffer, storePrefixBuffer]),
+			this._cache,
+		);
+		return subStore;
+	}
+
+	public async get(key: Buffer): Promise<Buffer> {
+		const prefixedKey = this._getKey(key);
+		const { value, deleted } = this._cache.get(prefixedKey);
+		if (value) {
+			return copyBuffer(value);
+		}
+		if (deleted) {
+			throw new NotFoundError(prefixedKey);
+		}
+		let persistedValue;
+		try {
+			persistedValue = await this._db.get(prefixedKey);
+		} catch (error) {
+			if (error instanceof DBNotFoundError) {
+				throw new NotFoundError(prefixedKey);
+			}
+			throw error;
+		}
+		this._cache.cache(prefixedKey, persistedValue);
+		return copyBuffer(persistedValue);
+	}
+
+	public async getWithSchema<T>(key: Buffer, schema: Schema): Promise<T> {
+		const value = await this.get(key);
+		return codec.decode<T>(schema, value);
+	}
 
-	public constructor(dataAccess: DataAccess, additionalInformation: AdditionalInformation) {
-		this.account = new AccountStore(dataAccess, {
-			defaultAccount: additionalInformation.defaultAccount,
+	public async set(key: Buffer, value: Buffer): Promise<void> {
+		const prefixedKey = this._getKey(key);
+		// 1. it does exist in cache just needs update => update
+		// 2. it did exist in cache, but it was deleted => update
+		if (this._cache.existAny(prefixedKey)) {
+			this._cache.set(prefixedKey, value);
+			return;
+		}
+		// 3. it does not exist in cache, but it does exist in DB => cache first and update
+		const dataExist = await this._ensureCache(prefixedKey);
+		if (dataExist) {
+			this._cache.set(prefixedKey, value);
+			return;
+		}
```

Comment:
I think 2 separate cache calls `cache.cache`,`cache.set` can be avoided.
```suggestion
		let dataExist: bool;
		try {
			const value = await this._db.get(prefixedKey);
			dataExist = true;
		} catch (error) {
			if (error instanceof DBNotFoundError) {
				dataExist = false;
			}
		}
		if (dataExist) {
			this._cache.cache(prefixedKey, value);
			return;
		}
```

Reply:
explicit `ensureCache` is important because if there is data in the database, we want to `cache` and then set the value.
`cache` will `cache` the value in `cacheDB` and also keep the `init` value from the database for the `diff` calculation.
so basically it's `cache`ing the original value in the database and then setting the new value from the input